[Woo POS][Cash & Receipts] Render cash button and navigate to cash view

iamgabrielma opened the pull request:
<!-- Remember about a good descriptive title. -->

Closes: #14603
Closes: #14599

## Description
This PR adds the `cash payment` button to the POS dashboard, and the initial navigation back and forth between both screens.

Please disregard the UI details on the cash view, we'll address the design when implementing the cash functionality and error handling.

## Changes
- Adds two new payment states: `.acceptingCash` and `.cashPaymentSuccessful`
- Adds a `PointOfSaleCollectCashView`
- As with Android, we present the view as standalone view, rendered from the POS dashboard. Upon tapping in cash collection we switch the payment state, which drives rendering the new view, upon navigating back, we return to the previous payment state which restores the dashboard as it was.

## Testing
- Enable `.acceptCashForPointOfSale` 
- Run the POS on a physical device
- Start to process a cart checkout, with a disconnected reader, observe the `cash payment` buttons appears and it's tappable below the reader not connected error:

![Simulator Screenshot - iPad Air 11 - iOS 17 5 M2 - 2024-12-19 at 17 19 10](https://github.com/user-attachments/assets/15b0a6f6-2351-4090-88d8-e0fffdc85c8b)

- Observe the order total appears on the the right top corner below the navigation button

<img width="232" alt="Screenshot 2024-12-19 at 17 20 06" src="https://github.com/user-attachments/assets/5eec57ee-6375-4203-9201-0f652b4e6395" />


- Connect the reader and observe the button still renders. Tapping the button and navigating back and forth retains the payment state, so the reader should remain connected:

https://github.com/user-attachments/assets/50b25ced-a2a8-4651-ad13-0c168a82c4ae

- At the moment "Mark payment as complete" is not functional.

## Screenshots
<!-- Include before and after images or gifs when appropriate. -->

---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

## Reviewer (or Author, in the case of optional code reviews):

Please make sure these conditions are met before approving the PR, or request changes if the PR needs improvement:

- [x] The PR is small and has a clear, single focus, or a valid explanation is provided in the description. If needed, please request to split it into smaller PRs.
- [x] Ensure Adequate Unit Test Coverage: The changes are reasonably covered by unit tests or an explanation is provided in the PR description.
- [x] Manual Testing: The author listed all the tests they ran, including smoke tests when needed (e.g., for refactorings). The reviewer confirmed that the PR works as expected on all devices (phone/tablet) and no regressions are added.

iamgabrielma committed:
Render cash button and navigate to cash view

dangermattic commented:
<!--
  0 Errors
  1 Warning: View files have been modified,...
  1 Message: This PR is still a Draft: some...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">View files have been modified, but no screenshot or video is included in the pull request. Consider adding some for clarity.</td>
    </tr>
  </tbody>
</table>
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">This PR is still a Draft: some checks will be skipped.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

iamgabrielma committed:
render totals in cash view navigation

iamgabrielma committed:
move strings to extension

iamgabrielma committed:
update button style

iamgabrielma commented:
👋 there's no rush to review this as targets 21.4, it can wait to next week.

joshheald commented on WooCommerce/Classes/POS/Models/PointOfSaleAggregateModel.swift:
```diff
@@ -61,6 +61,7 @@ class PointOfSaleAggregateModel: ObservableObject, PointOfSaleAggregateModelProt
 
     private var startPaymentOnCardReaderConnection: AnyCancellable?
     private var cardReaderDisconnection: AnyCancellable?
+    private var latestPaymentState: PointOfSalePaymentState? = nil
```
To avoid issues around keeping state up to date, it would be better to avoid this duplication. Can we get what we need from the `paymentState`?

iamgabrielma replied:
In this case you mean to keep track of "latest state before it switches to new state" internally within `PointOfSalePaymentState`, and then access that? It definitely feels better, yes. For the moment I've [reverted](71547ba0c3b96c291092f92e39ca790104346b39) the changes to payment state with the changes to navigation, but I'll take that into account going forward.

joshheald commented on WooCommerce/Classes/POS/Models/PointOfSaleAggregateModel.swift:
```diff
@@ -263,6 +264,23 @@ extension PointOfSaleAggregateModel {
+        if let latestPaymentState = latestPaymentState {
+            paymentState = latestPaymentState
+        }
+    }
```
I'm not sure that this is particularly safe. Setting the state back to something it used to be risks it being out-of-sync with the card payment service, and what happens then?

It would probably be better to cancel the reader preparation when they tap cash payment, then trigger it again if they cancel. I believe that's what's being done on Android too, so keeping platform consistency would be good.

iamgabrielma replied:
That makes sense, thanks for the pointer. I've reverted these changes for the time being.

> It would probably be better to cancel the reader preparation when they tap cash payment, then trigger it again if they cancel. I believe that's what's being done on Android too, so keeping platform consistency would be good.

👍 I've added a note to https://github.com/woocommerce/woocommerce-ios/issues/14682 for when implementing collecting and cancelling

joshheald commented on WooCommerce/Classes/POS/Models/PointOfSalePaymentState.swift:
```diff
@@ -49,7 +51,9 @@ extension PointOfSalePaymentState {
-                .cardPaymentSuccessful:
+                .cardPaymentSuccessful,
+                .acceptingCash,
+                .cashPaymentSuccessful:
```
Is the distinction between `card` and `cash` needed on the success state? If so, perhaps they should be distinguished using an associated value?

Should `acceptingCash` actually be a state here? Probably, but it's arguable.

iamgabrielma replied:
Good question. Some of the presentation logic for the rest of the collection flow checks if `.cardPaymentSuccessful`, which won't make much sense as state when the order is paid successfully via cash. Since moving the cash view presentation to not rely on payment state I think right now we can get rid of `.acceptingCash`, but we still need to do something with `.cardPaymentSuccessful`. 

I'll check both into using a new state, or as you mention making this a `.paymentSuccessful` state with associated value for cash or card 👍

joshheald commented on WooCommerce/Classes/POS/Presentation/PointOfSaleCollectCashView.swift:
```diff
@@ -0,0 +1,129 @@
+            return totals.orderTotal
+        }
+        return nil
+    }
```
Perhaps you can simplify this view by eliminating the potential for nil here?

If you required that `totals` gets passed in, then only allow this view to be presented when the orderState is loaded, you'd avoid it entirely.

iamgabrielma replied:
Oh definitely, updated here: https://github.com/woocommerce/woocommerce-ios/pull/14724/commits/714dfe5115cdeb8146a10cc4fa59eeb3b6da568a

joshheald commented on WooCommerce/Classes/POS/Presentation/PointOfSaleCollectCashView.swift:
```diff
@@ -0,0 +1,129 @@
+                    .foregroundColor(.red)
+            }
+
+            Button(action: {
+                Task { @MainActor in
+                    await markComplete()
+                }
+            }, label: {
+                HStack(spacing: Constants.buttonSpacing) {
+                    if isLoading {
+                        ProgressView()
+                            .progressViewStyle(CircularProgressViewStyle())
+                            .tint(Color.posPrimaryTextInverted)
+                    } else {
+                        Text(Localization.markPaymentCompletedButtonTitle)
+                            .font(Constants.buttonFont)
+                    }
+                }
+                .frame(maxWidth: .infinity)
+            })
```
Perhaps another place where an AsyncButton would be good? It'd need a binding so that the text field's `onSubmit` also triggered the progress view, but that seems like a useful option.

iamgabrielma replied:
Agree! I'll make the change with https://github.com/woocommerce/woocommerce-ios/issues/14602

joshheald commented on WooCommerce/Classes/POS/Presentation/PointOfSaleCollectCashView.swift:
```diff
@@ -0,0 +1,129 @@
+                }
+                .frame(maxWidth: .infinity)
+            })
+            .padding(Constants.buttonPadding)
+            .frame(maxWidth: .infinity)
+            .foregroundColor(Color.posPrimaryTextInverted)
+            .background(Color.posOverlayFillInverted)
+            .cornerRadius(Constants.buttonCornerRadius)
+            .contentShape(Rectangle())
```
I'm guessing some of these are shared with other buttons? It'd be good to put them in a button style for sharing and to make the view easier to read.

iamgabrielma replied:
Yes, we have an existing issue to make the async button a reusable component, I've logged it along: https://github.com/woocommerce/woocommerce-ios/issues/14626

joshheald commented on WooCommerce/Classes/POS/Presentation/PointOfSaleCollectCashView.swift:
```diff
@@ -0,0 +1,129 @@
+            comment: "Title of the cash payment navigation back button"
+        )
+        static let markPaymentCompletedButtonTitle = NSLocalizedString(
+            "pointOfSale.cashview.back.navigation.title",
```
This key is wrong 😬 copy-pasta from above

iamgabrielma replied:
Oops! Fixed: https://github.com/woocommerce/woocommerce-ios/pull/14724/commits/714dfe5115cdeb8146a10cc4fa59eeb3b6da568a

joshheald commented on WooCommerce/Classes/POS/Presentation/PointOfSaleCollectCashView.swift:
```diff
@@ -0,0 +1,129 @@
+        )
+        static let markPaymentCompletedButtonTitle = NSLocalizedString(
+            "pointOfSale.cashview.back.navigation.title",
+            value: "Mark payment as complete",
```
This is a weird way to word it... I can see it's in the designs, but maybe it's worth discussing some more? It feels very programmery, I think the merchant action is more like `Accept payment`

iamgabrielma replied:
Yeah, I'll raise it with Wagner along the navigation icons for each platform

joshheald commented on WooCommerce/Classes/POS/Presentation/PointOfSaleDashboardView.swift:
```diff
@@ -87,9 +87,14 @@ struct PointOfSaleDashboardView: View {
+                        TotalsView()
+                            .accessibilitySortPriority(2)
+                            .transition(.move(edge: .trailing))
+                    }
```
This feels like the wrong place for the views to be switched. The button to collect cash is on the TotalsView, it seems like that should also do the presentation, rather than bounce out to the dashboard.

I know TotalsView is big, but collecting cash still seems like part of its responsibilities, not the dashboard's.

iamgabrielma replied:
Sounds good, I've moved this to be a modal presentation from Totals: 6eb26efa0c7a615e7714c5541dc18c1a98e9a422

joshheald reviewed (commented):
Is it supposed to be modal, or a push navigation? It looks and feels like it's supposed to be a fullscreen cover modal, but it's not totally clear because of the back button. Certainly fullscreen cover would be a closer match to the bottom sheet shown in Wagner's walkthrough. 

A couple of areas to look at:

1. The transition to and from cash payments is weird, and I think it's because of the view being presented from the dashboard.
2. Payment state – as a general principle, don't mess with the source of truth, it'll cause bugs and make everything complicated!

## Transition weirdness
I don't think presenting cash payments from the dashboard is ideal – perhaps try a `fullscreenCover` over the Totals view if it should be modal. If you do this, use a close button not the back button.

Note the way the underlying totals views navigate in and out. Fixing this is going to be difficult with the way you're presenting it at the moment, I'd advise against it to keep things simple! A `fullscreenCover` presentation _should_ just avoid all these issues.

https://github.com/user-attachments/assets/bd14c674-5149-4a74-b313-1167900bfb37

## Payment state
The aggregate model _exposes_ the source of truth for (card) payments, but it comes from the card payment service – you can't just change it. If you do, the card payment service will keep doing its thing and that makes the cash payment go away.

When the merchant taps `Cash payment`, they are implicitly saying "This is not a card payment". So we should cancel the card payment preparation, otherwise the reader will just be sitting there waiting to be tapped. The simulator shows this really well because it insta-taps.

Fortunately we already support cancellations and re-preparation well: we do it when you go back to the item list and forward to the checkout again. Those functions should be all you need to handle the state management for cash payments.

https://github.com/user-attachments/assets/71221689-771b-4699-ab1c-445f608ced01


Various other comments inline. I know the UI's not done yet, but just a note – that back arrow is very Androidy, and perhaps these designs are Android-first so need some translation. I think we should use a `chevron.backward` where we need similar for iOS, but you could raise it with Wagner to check. This feels like somewhere platform convention comes before the design system.

iamgabrielma committed:
Make cash view modal presentation from totals

iamgabrielma committed:
Pass order total to cash view from parent

iamgabrielma committed:
update backbutton icon on receipt for consistency

Along with the usage of a chevron for the cash view when navigating back, we also update the receipt view with the same icon for consistency

iamgabrielma committed:
Revert changes to payment state for cash navigation

iamgabrielma committed:
Render button only when payment state idle

iamgabrielma commented:
Thanks for the review and pointers @joshheald , very much appreciated.

> Is it supposed to be modal, or a push navigation? It looks and feels like it's supposed to be a fullscreen cover modal, but it's not totally clear because of the back button. Certainly fullscreen cover would be a closer match to the bottom sheet shown in Wagner's walkthrough

That's a good question, I think there's a bit of mix here, it looks like push navigation (from the Android demo this also seems to be pushed from the side) but then the design demo shows a modal. I've make the changes to present it from the totals view as a full screen modal and definitely feels much better and less prone to error. 

> The aggregate model exposes the source of truth for (card) payments, but it comes from the card payment service – you can't just change it. If you do, the card payment service will keep doing its thing and that makes the cash payment go away.

I wasn't feeling very good about this, and I guess the gut feeling was right 😅 I've removed the navigation interaction with the payment state for the time being. I've also updated the [conditions](5f9bc07803b60970d6f0cfc299c677564b2522bb) when the button should be rendered so it only happens when the payment state is .idle for now. I'll need to test this further when playing with reader connection/cancellation:


https://github.com/user-attachments/assets/886e31be-fc9e-4add-8b60-1f48f9800d13


> that back arrow is very Androidy, and perhaps these designs are Android-first so need some translation. I think we should use a chevron.backward where we need similar for iOS

Yes, I've updated the arrow to a chevron (both in cash and receipt views) for consistency, I'll bring this up with design in any case, along with the string recommendations.

Changes:

✅ Present the view as a full-screen modal from totals
✅ Pass order totals values from the paren
✅ Updated navigation style to be more iOS
✅ Removed changes to payment state
✅ Added additional conditions for rendering the cash button

iamgabrielma committed:
Merge branch 'trunk' into task/14603-add-cash-option-to-totals-view

jaclync commented on WooCommerce/Classes/POS/Presentation/TotalsView.swift:
```diff
@@ -70,6 +89,12 @@ struct TotalsView: View {
         .geometryGroupIfSupported()
+        .fullScreenCover(isPresented: $shouldShowCollectCashPayment) {
+            if case .loaded(let total) = posModel.orderState {
+                PointOfSaleCollectCashView(isVisible: $shouldShowCollectCashPayment,
```
❓ It looks like `isVisible` is only used for dismissing the `PointOfSaleCollectCashView` fullscreen cover, can the view be dismissed with the `@Environment(\.dismiss) private var dismiss`? If it's really needed, just a dismiss closure would be more clear than `isVisible` parameter since `fullScreenCover(isPresented: $shouldShowCollectCashPayment)` already uses the binding to present the view.

iamgabrielma replied:
Ah good catch, that was my initial approach but didn’t work correctly as we were pushing navigation into the stack, dismissing the whole POS, but since we’ve moved to a modal presentation instead it's much simpler to use dismiss. Changed on https://github.com/woocommerce/woocommerce-ios/pull/14724/commits/8d723fa9b9002f3d500150efc66990a917d2925c

jaclync commented on WooCommerce/Classes/POS/ViewHelpers/CartViewHelper.swift:
```diff
@@ -32,7 +32,9 @@ private extension PointOfSalePaymentState {
-        case .idle, .validatingOrderError, .acceptingCard:
+        case .idle,
+                .validatingOrderError,
+                .acceptingCard:
```
nit: it doesn't look like there are any changes, coulud be reverted.

iamgabrielma replied:
Reverted! https://github.com/woocommerce/woocommerce-ios/pull/14724/commits/e656b3b3f4db2f5d20219096b307bc537e85aac2

jaclync commented on WooCommerce/Classes/POS/Presentation/TotalsView.swift:
```diff
@@ -291,6 +316,8 @@ private extension TotalsView {
     enum Constants {
         static let pricesIdealWidth: CGFloat = 382
         static let verticalSpacing: CGFloat = 56
+        static let buttonHeight: CGFloat = 56
```
nit: fixed button height might result in text being clipped with a large dynamic font size, and/or multiline text. If a fixed height is needed, `@ScaledMetric` can be considered.

iamgabrielma replied:
I attempted to make the changes here but I found several other issues through the view when using dynamic font sizes, so I'll tackle these separately, as an example some additional space is starting to show up at the top of the view:
![Simulator Screenshot - iPad Air 11 - iOS 17 5 M2 - 2024-12-23 at 18 56 25](https://github.com/user-attachments/assets/e96aeee5-6160-4e10-94d9-5281ddfe4084)

jaclync commented on WooCommerce/Classes/POS/Presentation/PointOfSaleCollectCashView.swift:
```diff
@@ -0,0 +1,137 @@
+    @State private var isLoading: Bool = false
+    @State private var errorMessage: String?
+
+    @Binding var isVisible: Bool
```
nit: does this need to be public? (If this binding is still needed, from another comment in `TotalsView`)

iamgabrielma replied:
No need! Removed since we're using dismiss now

jaclync reviewed (commented):
I will finish the review with a physical card reader in a few hours, posting the initial review first.

In dark mode, iPad 10th gen iOS 18.0 simulator, I noticed some unexpected behavior from the cart view **only when the feature flag is enabled** as in the screencast below:
- When navigating to the totals screen, the top bar in the cart view has a different background color
- After dismissing the cash payment view, the cart view also gets shifted vertically briefly

This cart view behavior does not take place when the feature flag is disabled or in trunk.


https://github.com/user-attachments/assets/a568d34b-c865-49b4-a584-a3d72e44202b

Not sure if it's from some animation or some part of presentation of the cash payment view?

jaclync reviewed (commented):
Just tested on a physical tablet and this step didn't work on my end:

> Connect the reader and observe the button still renders. Tapping the button and navigating back and forth retains the payment state, so the reader should remain connected:

After connecting the reader, the Cash payment CTA is not shown anymore:


https://github.com/user-attachments/assets/1e734515-1ed3-4f6d-99ea-14f9188e45e7

This doesn't seem expected from the screencast for this step in the PR description?

iamgabrielma committed:
revert styling

iamgabrielma committed:
Update cash view to dismiss via environment

iamgabrielma committed:
Resolved animation oddness

When dismissing the cash view, the animation for cart seems to jump vertically. It seems to be resolved by adding the cash view to matched geometry

iamgabrielma committed:
Show button when card is acceptingCard state

iamgabrielma commented:
Thanks for the review and testing @jaclync !

> When navigating to the totals screen, the top bar in the cart view has a different background color

I'm unable to replicate this, now in iOS17 nor 18 🤔 do you see it in the physical device as well?

> After dismissing the cash payment view, the cart view also gets shifted vertically briefly

I see the same, the cart definitely seems to jump when dismissing. I fixed it on https://github.com/woocommerce/woocommerce-ios/pull/14724/commits/c8a00b29da8baca25bbf4a01628a572415e7310b by using the matching geometry. That said I'm seeing some oddness in the simulator that does not happen in physical device, I dropped a comment in slack here: p1734945310025859-C025A8VV728

> After connecting the reader, the Cash payment CTA is not shown anymore [...] This doesn't seem expected from the screencast for this step in the PR description?

Good catch, yes and no. My fault as I didn't update the instructions: We don't want to retain the state anymore, but to cancel the payment intent if the operator decides to use cash (this will be implemented later), but you're right that the cash payment button should appear when we're accepting card but haven't tapped yet. I've addressed this here: [57a66b6](https://github.com/woocommerce/woocommerce-ios/pull/14724/commits/57a66b6c50bfa30eda51d90e2e989602d8c5441b)


![20241224 show button when correct paymentstate](https://github.com/user-attachments/assets/4ffeccd5-de31-44ad-b436-c9da4bf483db)

jaclync reviewed (approved):
The cash payment CTA is shown when the reader is ready now 👍 

It's interesting I don't see the cart view header background color change in iOS 17.5.1 iPad device, only in one iOS 18.0 iPad 10th gen simulator. In another iOS 18.0 iPad Pro 11in simulator, the background color only changed briefly when the vertical move happened after dismissing the cash view 😅 

Since this doesn't happen when the feature flag is disabled, we can try testing on an iOS 18 physical tablet to confirm outside of this PR. I can upgrade my iPad to iOS 18 if needed.


iOS 17.5.1 device | iOS 18.0 simulator
-- | --
![IMG_0448](https://github.com/user-attachments/assets/feb0246e-ba2d-441a-bd8f-64450a4f4bc1) | ![Simulator Screenshot - iPad (10th generation) - 2024-12-24 at 09 02 50](https://github.com/user-attachments/assets/f9709bec-3d5f-4f0e-8376-60db693dd139)

iamgabrielma commented:
> Since this doesn't happen when the feature flag is disabled, we can try testing on an iOS 18 physical tablet to confirm outside of this PR. I can upgrade my iPad to iOS 18 if needed.

Thanks for the screenshots! Let's upgrade either yours or mine so we still keep one in iOS17 for testing. I'll drop a follow-up in Slack 👍

mokagio commented:
@iamgabrielma as you reported internally, p1734920797831869-slack-CC7L49W13, the prototype build fails. Buried. inthe logs, I found these compilation errors

```
❌ /opt/ci/builds/builder/automattic/woocommerce-ios/WooCommerce/Classes/POS/Presentation/PointOfSaleCollectCashView.swift:131:26: cannot find 'PointOfSalePreviewItemsController' in scope
        itemsController: PointOfSalePreviewItemsController(),
                         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
❌ /opt/ci/builds/builder/automattic/woocommerce-ios/WooCommerce/Classes/POS/Presentation/PointOfSaleCollectCashView.swift:132:36: cannot find 'CardPresentPaymentPreviewService' in scope
        cardPresentPaymentService: CardPresentPaymentPreviewService(),
                                   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
❌ /opt/ci/builds/builder/automattic/woocommerce-ios/WooCommerce/Classes/POS/Presentation/PointOfSaleCollectCashView.swift:133:26: cannot find 'PointOfSalePreviewOrderController' in scope
        orderController: PointOfSalePreviewOrderController())
                         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

https://buildkite.com/automattic/woocommerce-ios/builds/27034#0193f136-9e0e-4804-ae58-1c83a6164959/735-8363

iamgabrielma committed:
Merge branch 'trunk' of https://github.com/woocommerce/woocommerce-ios into task/14603-add-cash-option-to-totals-view

iamgabrielma committed:
Wrap preview in debug precompiler flag
